server/client_test.go

```diff
@@ -485,6 +486,175 @@ func TestClientPubWithQueueSub(t *testing.T) {
+			perms:   &SubjectPermission{Allow: []string{"> bar.>"}},
+			subject: "whizz",
+			want:    "-ERR 'Permissions Violation for Subscription to \"whizz\"'\r\n",
+		},
+		{
+			name:    "deny plain subscription on foo",
+			perms:   &SubjectPermission{Allow: []string{">"}, Deny: []string{"foo"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"bar\"'\r\n",
+		},
+		{
+			name:    "allow plain subscription, except foo",
+			perms:   &SubjectPermission{Allow: []string{">"}, Deny: []string{"foo"}},
+			subject: "bar",
+			want:    "+OK\r\n",
+		},
+		{
+			name:    "deny everything",
+			perms:   &SubjectPermission{Allow: []string{">"}, Deny: []string{">"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"bar\"'\r\n",
+		},
+		{
+			name:    "can only subscribe to queues v1",
+			perms:   &SubjectPermission{Allow: []string{"> v1.>"}},
+			subject: "foo",
+			queue:   "v1.prod",
+			want:    "+OK\r\n",
+		},
+		{
+			name:    "cannot subscribe to queues, plain subscribe ok",
+			perms:   &SubjectPermission{Allow: []string{">"}, Deny: []string{"> >"}},
+			subject: "foo",
+			want:    "+OK\r\n",
+		},
+		{
+			name:    "cannot subscribe to queues, queue subscribe not ok",
+			perms:   &SubjectPermission{Deny: []string{"> >"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"bar\"'\r\n",
+		},
+		{
+			name:    "deny all queue subscriptions on dev or stg only",
+			perms:   &SubjectPermission{Deny: []string{"> *.dev", "> *.stg"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "+OK\r\n",
+		},
+		{
+			name:    "allow only queue subscription on dev or stg",
+			perms:   &SubjectPermission{Allow: []string{"> *.dev", "> *.stg"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"bar\"'\r\n",
+		},
+		{
+			name:    "deny queue subscriptions with subject foo",
+			perms:   &SubjectPermission{Deny: []string{"foo >"}},
+			subject: "foo",
+			queue:   "bar",
+			want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"bar\"'\r\n",
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			_, client, r := setupClient()
+
+			client.RegisterUser(&User{
+				Permissions: &Permissions{Subscribe: c.perms},
+			})
+			connect := []byte("CONNECT {\"verbose\":true}\r\n")
+			qsub := []byte(fmt.Sprintf("SUB %s %s 1\r\n", c.subject, c.queue))
+
+			go client.parseFlushAndClose(append(connect, qsub...))
+
+			var buf bytes.Buffer
+			if _, err := io.Copy(&buf, r); err != nil {
+				t.Fatal(err)
+			}
+
+			// Extra OK is from the successful CONNECT.
+			want := "+OK\r\n" + c.want
+			if got := buf.String(); got != want {
+				t.Fatalf("Expected to receive %q, but instead received %q", want, got)
+			}
+		})
+	}
+}
+
```

Comment:
I think we need to add a test (simpler I think that adding to existing test) where you have a the presence of a regular sub and try to queue subscribe. We want to check the conditions where say plain sub is allowed, but you try to queue subscribe with a queue name that is not in the list (or there is no queue name in the list), etc...

Reply:
I appended two more cases to our list. Does this cover what you were thinking?

```
{
	name:    "plain sub is allowed, but queue subscribe with queue not in list",
	perms:   &SubjectPermission{Allow: []string{"foo bar"}},
	subject: "foo",
	queue:   "fizz",
	want:    "-ERR 'Permissions Violation for Subscription to \"foo\" using queue \"fizz\"'\r\n",
},
{
	name:    "allow plain sub, but do queue subscribe",
	perms:   &SubjectPermission{Allow: []string{"foo"}},
	subject: "foo",
	queue:   "bar",
	want:    "+OK\r\n",
},
```

Reply:
Since you create only a queue sub, the last test should expect an error, no?

Reply:
Oh, I take that back. As long as no queue name is specified, it means that all queues are allowed.